src/tdastro/math_nodes/given_sampler.py

```diff
@@ -73,42 +77,108 @@ def compute(self, graph_state, rng_info=None, **kwargs):
+        -------
+        results : any
+            The result of the computation. This return value is provided so that testing
+            functions can easily access the results.
+        """
+        rng = rng_info if rng_info is not None else self._rng
+
+        if graph_state.num_samples == 1:
+            inds = rng.integers(0, self._num_values)
+        else:
+            inds = rng.integers(0, self._num_values, size=graph_state.num_samples)
+
+        return self.values[inds]
+
+
+class TableSampler(NumpyRandomFunc):
+    """A FunctionNode that returns values from a table-like data,
+    including a Pandas DataFrame or AstroPy Table. The results returned
+    can be in-order (for testing) or randomly selected with replacement.
+
+    Parameters
+    ----------
     data : pandas.DataFrame, astropy.table.Table, or dict
         The object containing the data to sample.
+    in_order : bool
+        Return the given data in order of the rows (True). If False, performs
+        random sampling with replacement. Default: False
+
+    Attributes
+    ----------
+    data : astropy.table.Table
+        The object containing the data to sample.
     columns : list of str
         The column names for the output columns.
+    in_order : bool
+        Return the given data in order of the rows (True). If False, performs
+        random sampling with replacement. Default: False
     next_ind : int
-        The next index to sample.
+        The next index to sample for in order sampling.
+    num_values : int
+        The total number of items from which to draw the data.
     """
 
-    def __init__(self, data, node_label=None, **kwargs):
+    def __init__(self, data, in_order=False, **kwargs):
         self.next_ind = 0
+        self.in_order = in_order
 
         if isinstance(data, dict):
-            self.data = pd.DataFrame(data)
+            self.data = Table(data)
         elif isinstance(data, Table):
-            self.data = data.to_pandas()
-        elif isinstance(data, pd.DataFrame):
             self.data = data.copy()
+        elif isinstance(data, pd.DataFrame):
+            self.data = Table.from_pandas(data)
         else:
             raise TypeError("Unsupported data type for TableSampler.")
 
-        # Add each of the flow's data columns as an output parameter.
-        self.columns = [x for x in self.data.columns]
-        super().__init__(self._non_func, node_label=node_label, outputs=self.columns, **kwargs)
+        # Check there are some rows.
+        self._num_values = len(self.data)
+        if self._num_values == 0:
+            raise ValueError("No data provided to TableSampler.")
 
-    def _non_func(self):
-        """This function does nothing. Everything happens in the overloaded compute()."""
-        pass
+        # Add each of the flow's data columns as an output parameter.
+        self.columns = [x for x in self.data.colnames]
+        super().__init__("uniform", outputs=self.columns, **kwargs)
```

Comment:
I don't think we need a new list here 
```suggestion
        super().__init__("uniform", outputs=self.data.colnames, **kwargs)
```